packages/SwingSet/src/controller.js

```diff
@@ -234,39 +259,135 @@ export async function buildVatController(config, argv = []) {
+      );
+    } else if (count === 0) {
+      throw Error(
+        `config ${groupName}.${descName}: you must specify one of: "bundleName", "bundle", "bundlePath", or "sourcePath"`,
+      );
+    }
+    if (kernel.hasBundle(descName) && !desc.bundleName) {
+      throw Error(`config ${groupName}: bundle ${descName} multiply defined`);
+    }
+  }
+
+  async function bundleBundles(group, groupName) {
+    if (group) {
+      const names = [];
+      const presumptiveBundles = [];
+      for (const name of Object.keys(group)) {
+        const desc = group[name];
+        validateBundleDescriptor(desc, groupName, name);
+        if (!desc.bundleName) {
+          names.push(name);
+          if (desc.sourcePath) {
+            const sourcePath = desc.sourcePath;
+            if (!(sourcePath[0] === '.' || path.isAbsolute(sourcePath))) {
+              throw Error(
+                'sourcePath must be relative (./foo) or absolute (/foo) not bare (foo)',
+              );
+            }
+            presumptiveBundles.push(bundleSource(sourcePath));
+          } else if (desc.bundlePath) {
+            const bundlePath = desc.bundlePath;
+            if (!(bundlePath[0] === '.' || path.isAbsolute(bundlePath))) {
+              throw Error(
+                'bundlePath must be relative (./foo) or absolute (/foo) not bare (foo)',
+              );
+            }
+            presumptiveBundles.push(fs.readFileSync(bundlePath));
+          } else if (desc.bundle) {
+            presumptiveBundles.push(desc.bundle);
+          } else {
+            assert.fail(`this can't happen`);
+          }
+        }
+      }
+      const actualBundles = await Promise.all(presumptiveBundles);
+      for (let i = 0; i < names.length; i += 1) {
+        kernel.addBundle(names[i], actualBundles[i]);
+      }
+    }
   }
 
-  if (config.vats) {
-    for (const name of config.vats.keys()) {
-      const v = config.vats.get(name);
-      // eslint-disable-next-line no-await-in-loop
-      await addGenesisVat(name, v.sourcepath, v.options || {});
+  await bundleBundles(config.bundles, 'bundles');
+  await bundleBundles(config.vats, 'vats');
+  // await bundleBundles(config.devices, 'devices'); // TODO: refactor device config
+
+  if (config.devices) {
+    const devices = [];
+    for (const [name, srcpath, endowments] of config.devices) {
+      devices.push(addGenesisDevice(name, srcpath, endowments));
     }
+    await Promise.all(devices);
   }
 
-  let bootstrapVatName;
-  if (config.bootstrapIndexJS) {
-    bootstrapVatName = '_bootstrap';
-    await addGenesisVat(bootstrapVatName, config.bootstrapIndexJS);
+  if (config.vats) {
+    const vats = [];
+    for (const name of Object.keys(config.vats)) {
+      const v = config.vats[name];
+      vats.push(
+        addGenesisVat(
+          name,
+          v.bundleName || name,
+          v.options || {},
```

Comment:
I'm thinking that `parameters` may be better, simply because we tend to use `options` or `blahblahblahOptions` as a final parameter in a function's args, and although that's what this translates to internally (hence my thinking when I used the word "options" here), from the perspective of somebody authoring a config file or composing a config object that's not how they use it.  If we want to make this change I'm actually inclined to do it now, since it just revises a bunch of files that are already touched in this PR to touch them slightly different; it won't affect the size of the PR itself of the volume of stuff to be reviewed significantly if at all.